src/uu/mknod/src/mknod.rs

```diff
@@ -31,168 +50,168 @@ fn makedev(maj: u64, min: u64) -> dev_t {
-                    return 1;
-                }
-                newmode = parsed;
-            }
-            Err(e) => {
-                show_info!("{}", e);
-                return 1;
-            }
+        // set umask back to original value
+        if set_umask {
+            libc::umask(last_umask);
         }
-        unsafe {
-            last_umask = libc::umask(0);
+
+        if errno == -1 {
+            let c_str = CString::new(NAME).expect("Failed to convert to CString");
+            // shows the error from the mknod syscall
+            libc::perror(c_str.as_ptr());
         }
+        errno
     }
+}
 
-    let mut ret = 0i32;
-    match matches.free.len() {
-        0 => show_usage_error!("missing operand"),
-        1 => show_usage_error!("missing operand after ‘{}’", matches.free[0]),
-        _ => {
-            let args = &matches.free;
-            let c_str = CString::new(args[0].as_str()).expect("Failed to convert to CString");
-
-            // Only check the first character, to allow mnemonic usage like
-            // 'mknod /dev/rst0 character 18 0'.
-            let ch = args[1]
-                .chars()
-                .next()
-                .expect("Failed to get the first char");
-
-            if ch == 'p' {
-                if args.len() > 2 {
-                    show_info!("{}: extra operand ‘{}’", NAME, args[2]);
-                    if args.len() == 4 {
-                        eprintln!("Fifos do not have major and minor device numbers.");
-                    }
-                    eprintln!("Try '{} --help' for more information.", NAME);
-                    return 1;
-                }
+pub fn uumain(args: impl uucore::Args) -> i32 {
+    // Linux-specific options, not implemented
+    // opts.optflag("Z", "", "set the SELinux security context to default type");
+    // opts.optopt("", "context", "like -Z, or if CTX is specified then set the SELinux or SMACK security context to CTX");
 
-                ret = _makenod(c_str, S_IFIFO | newmode, 0);
-            } else {
-                if args.len() < 4 {
-                    show_info!("missing operand after ‘{}’", args[args.len() - 1]);
-                    if args.len() == 2 {
-                        eprintln!("Special files require major and minor device numbers.");
-                    }
-                    eprintln!("Try '{} --help' for more information.", NAME);
-                    return 1;
-                } else if args.len() > 4 {
-                    show_usage_error!("extra operand ‘{}’", args[4]);
-                    return 1;
-                } else if !"bcu".contains(ch) {
-                    show_usage_error!("invalid device type ‘{}’", args[1]);
-                    return 1;
-                }
+    let matches = App::new(executable!())
+        .version(VERSION)
+        .usage(USAGE)
+        .after_help(LONG_HELP)
+        .about(ABOUT)
+        .arg(
+            Arg::with_name("mode")
```

Comment:
For consistency, please create an option structure like in
https://github.com/uutils/coreutils/blob/master/src/uu/chown/src/chown.rs#L30-L34
and use it in the rest of the program

Reply:
Do you mean for all arguments or just for `mode`? If you mean just for `mode`, I have some doubts if this is really helpful, because the values are something like `a-w` `644` etc and I'm just passing it on to the `mode` module which again expects a string. I think the structure would have to be quite involved to accurately capture any value it can have (unless of course you expect it to be only a String wrapper, but in that case I also doubt the value :)

Reply:
Nope, creating a structure to store the various options :)

See:
https://github.com/uutils/coreutils/search?q=%22pub+mod+options%22

Reply:
ping ?

Reply:
I didn't forget this. I'm just quite busy otherwise right now. I'll work on it asap.